Document/0x05f-Testing-Local-Authentication.md

```diff
@@ -82,21 +84,66 @@ Make sure that the unlocked key is used during the application flow. For example
+This is different to the `FingerprintManager` class that provided no UI and developers needed to build their own fingerprint UI and only fingerprint sensors are supported.
+
+A very detailed overview and explanation of the Biometric API on Android was published on the [Android Developer Blog](https://android-developers.googleblog.com/2019/10/one-biometric-api-over-all-android.html "One Biometric API Over all Android").
+
+##### FingerprintManager (deprecated in Android P)
+
+Android 6.0 (API level 23) introduced public APIs for authenticating users via fingerprint, but is deprecated in Android 9 (API level 28). Access to the fingerprint hardware is provided through the [FingerprintManager class](https://developer.android.com/reference/android/hardware/fingerprint/ "FingerprintManager"). An app can request fingerprint authentication by instantiating a `FingerprintManager` object and calling its `authenticate` method. The caller registers callback methods to handle possible outcomes of the authentication process (i.e. success, failure, or error). Note that this method doesn't constitute strong proof that fingerprint authentication has actually been performed - for example, the authentication step could be patched out by an attacker, or the "success" callback could be overloaded using dynamic instrumentation.
+
+Better security is achieved by using the fingerprint API in conjunction with the Android `KeyGenerator` class. With this method, a symmetric key is stored in the `KeyStore` and "unlocked" with the user's fingerprint. For example, to enable user access to a remote service, an AES key is created which encrypts the authentication token. By calling `setUserAuthenticationRequired(true)` when creating the key, it is ensured that the user must re-authenticate to retrieve it. The encrypted authentication token can then be saved directly on the the device (e.g. via `SharedPreferences`). This design is a relatively safe way to ensure the user actually entered an authorized fingerprint. Note however that this setup requires the app to hold the symmetric key in memory during cryptographic operations, potentially exposing it to attackers that manage to access the app's memory during runtime.

```

Comment:
- first sentence turned active.
- changed to Shared Preferences because we refer to it here a a storage method and not directly the class right. It could also be. But I feel this more appropriate, what do you think? 

```suggestion
You can achieve better security by using the fingerprint API in conjunction with the Android `KeyGenerator` class. With this approach, a symmetric key is stored in the Android KeyStore and unlocked with the user's fingerprint. For example, to enable user access to a remote service, an AES key is created which encrypts the authentication token. By calling `setUserAuthenticationRequired(true)` when creating the key, it is ensured that the user must re-authenticate to retrieve it. The encrypted authentication token can then be saved directly on the the device (e.g. via Shared Preferences). This design is a relatively safe way to ensure the user actually entered an authorized fingerprint. However, note that this setup requires the app to hold the symmetric key in memory during cryptographic operations, potentially exposing it to attackers that manage to access the app's memory during runtime.
```

Reply:
Question: Is it the "symmetric key" or the "authentication token" that is exposed in memory? It is not that all keys from the KeyStore never leave it even when performing crypto operations. However the result of decrypting the "authentication token" will be exposed in memory, right?

Reply:
You are right, good spot! Will update this accordingly

Reply:
I removed the sentences completely. Keys in the Keystore are not exportable https://developer.android.com/training/articles/keystore

See commit: https://github.com/OWASP/owasp-mstg/pull/1569/commits/5c56dfefccab06624f9acc02574d0f803fe50d8b